heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -3894,6 +4023,145 @@ introLLVMFieldContentsM ::
+  ImplM vars s r (ps :> LLVMPointerType w :> LLVMPointerType w)
+  (ps :> LLVMPointerType w :> LLVMPointerType w)
+  (Maybe (BV.BV sz1, BV.BV sz2))
+implLLVMFieldTryProveWordEq2 x fp1 fp2 =
+  implLLVMFieldTryProveWordEq x fp2 >>>= \case
+  Nothing ->
+    let fp2_true = llvmFieldSetTrue fp2 fp2 in
+    implSwapM x (ValPerm_LLVMField fp1) x (ValPerm_LLVMField fp2_true) >>>
+    implLLVMFieldSetTrue x fp1 >>>
+    let fp1_true = llvmFieldSetTrue fp1 fp1 in
+    implSwapM x (ValPerm_LLVMField fp2_true) x (ValPerm_LLVMField fp1_true) >>>
+    return Nothing
+  Just bv2 ->
+    let fp2' = llvmFieldSetEqWord fp2 bv2 in
+    implSwapM x (ValPerm_LLVMField fp1) x (ValPerm_LLVMField fp2') >>>
+    implLLVMFieldTryProveWordEq x fp1 >>>= \case
+      Nothing ->
+        let fp1_true = llvmFieldSetTrue fp1 fp1 in
+        implSwapM x (ValPerm_LLVMField fp2') x (ValPerm_LLVMField fp1_true) >>>
+        implLLVMFieldSetTrue x fp2' >>>
+        return Nothing
+      Just bv1 ->
+        let fp1' = llvmFieldSetEqWord fp2 bv1 in
+        implSwapM x (ValPerm_LLVMField fp2') x (ValPerm_LLVMField fp1') >>>
+        return (Just (bv1, bv2))
+
+-- | Attempt to split a pointer permission @ptr((rw,off,sz) |-> p)@ on top of
+-- the stack into two permissions of the form @ptr((rw,off,8*len) |-> p1)@ and
+-- @ptr((rw,off+len,sz-(8*len)) |-> p2)@, that is, into one field of size @len@
+-- bytes and one field of the remaining size. If @p@ can be coerced to an
+-- equality permission @eq(llvmword(bv))@ for a known constant bitvector @bv@,
+-- then @p1@ and @p2@ are equalities to the split of @bv@ into known smaller
+-- bitvectors, and otherwise they are both @true@.
+implLLVMFieldSplit ::
+  (NuMatchingAny1 r, 1 <= w, KnownNat w, 1 <= sz, KnownNat sz) =>
+  ExprVar (LLVMPointerType w) -> LLVMFieldPerm w sz -> Integer ->
+  ImplM vars s r (ps :> LLVMPointerType w :> LLVMPointerType w)
+  (ps :> LLVMPointerType w)
+  (AtomicPerm (LLVMPointerType w), AtomicPerm (LLVMPointerType w))
+implLLVMFieldSplit x fp sz_bytes
+  | Just (Some sz) <- someNat (sz_bytes * 8)
+  , Just fp_m_sz <- subNat' (llvmFieldSize fp) sz
+  , Left LeqProof <- decideLeq sz (llvmFieldSize fp)
+  , Left LeqProof <- decideLeq (knownNat @1) sz
+  , Left LeqProof <- decideLeq (knownNat @1) fp_m_sz =
+    withKnownNat sz $ withKnownNat fp_m_sz $
+    use implStateEndianness >>>= \endianness ->
+    implLLVMFieldTryProveWordEq x fp >>>= \case
+      Just bv
+        | Just (bv1, bv2) <- bvSplit endianness sz bv ->
+          implSimplM Proxy (SImpl_SplitLLVMWordField
+                            x (llvmFieldSetEqWord fp bv) bv1 bv2 endianness) >>>
+          return (Perm_LLVMField (llvmFieldSetEqWord fp bv1),
+                  Perm_LLVMField (llvmFieldAddOffsetInt
+                                  (llvmFieldSetEqWord fp bv2)
+                                  sz_bytes))
+      Just _ ->
+        error "implLLVMFieldSplit: unreachable case"
```

Comment:
Maybe it's worth mentioning here why this case is unreachable, because it's not immediately obvious to me.

Reply:
Ok, I added a comment about this